File: jetway-api/schemas/validations.py
from schemas.airline import FlightOut
from schemas.admins import AdminOut
from schemas.users import UserOutAdmin, UserOut
from schemas.customer import CustomerOut
from schemas.airline import AirlineOut, TicketOut
from sql_app import countries, flights

def flight_out(flight):
    origin = countries.get_country_by_id(flight.origin_country_id)
    destination = countries.get_country_by_id(flight.destination_country_id)
    flight_out = FlightOut(
        id=flight.id,
        airline=flight.airline.name, 
        origin=origin.country_name,
        destination=destination.country_name,
        departure=flight.departure_time,
        landing=flight.landing_time,
        remaining_tickets=flight.remaining_tickets
    )
    return flight_out

def admin_out(admin):
    admin_out = AdminOut(
        id=admin.id,
        user_id=admin.user_id,
        username=admin.user.username,
        first_name=admin.first_name,
        last_name=admin.last_name
    )
    return admin_out

def user_out_admin(user):
    user_out = UserOutAdmin(
        id=user.id,
        username=user.username,
        email=user.email,
        role_name=user.role.role_name
    )
    return user_out

def user_out(user):
    user_out = UserOut(
        id=user.id,
        username=user.username,
        email=user.email
    )
    return user_out

def customer_out(customer):
    customer_out = CustomerOut(
        id=customer.id,
        user_id=customer.user_id,
        username=customer.user.username, 
        first_name=customer.first_name,
        last_name=customer.last_name,
        address=customer.address, 
        phone_no=customer.phone_no,
        credit_card=customer.credit_card
    )
    return customer_out

def airline_out(airline):
    airline_out = AirlineOut(
        id=airline.id,
        user_id=airline.user_id,
        username=airline.user.username,
        name=airline.name,
        country=airline.country.country_name
    )
    return airline_out

def ticket_out(ticket):
    flight_db = flights.get_flight_by_id(ticket.flight_id)
    flight = flight_out(flight_db)
    ticket_out = TicketOut(
        id=ticket.id,
        flight_id=flight.id,
        airline=flight.airline,
        origin=flight.origin,
        destination=flight.destination,
        departure=flight.departure
    )
    return ticket_out

File: jetway-ui/src/services/validations.js
import * as yup from 'yup'
import moment from 'moment';

// Validation schemas used in various forms...

// Phone regex
const phoneRegex = /^[\+]?[(]?[0-9]{3}[)]?[-\s\.]?[0-9]{3}[-\s\.]?[0-9]{4,6}$/

// credit card regex
const creditCardRegex = /(^4[0-9]{12}(?:[0-9]{3})?$)|(^(?:5[1-5][0-9]{2}|222[1-9]|22[3-9][0-9]|2[3-6][0-9]{2}|27[01][0-9]|2720)[0-9]{12}$)|(3[47][0-9]{13})|(^3(?:0[0-5]|[68][0-9])[0-9]{11}$)|(^6(?:011|5[0-9]{2})[0-9]{12}$)|(^(?:2131|1800|35\d{3})\d{11}$)/

// login form validation schema.
export const loginSchema = yup.object().shape({
    username: yup.string().required(), 
    password: yup.string().min(4).max(15).required()
});

// Flight search validation schema.
export const flightSearchSchema = yup.object().shape({
    origin: yup.string().required("Please select origin country."),
    destination: yup.string()
        .test('isInternational', 'Currently only international flights supported.', (value) => {
        return value !== origin;
    })
    .required("Please select destination country."),
    departure: yup.date().min(moment().toISOString(), "You cannot search past flights.").required('Valid date is required.'),
});

// New airline validation schema.
export const airlineSchema = yup.object().shape({
    name: yup.string().required("Airline name required."),
    country: yup.string().required("Please select country.")
});

// New customer validation schema.
export const customerSchema = yup.object().shape({
    first_name: yup.string().required(),
    last_name: yup.string().required(),
    address: yup.string().required(),
    phone_no: yup.string().matches(phoneRegex, 'Invalid phone number').required(),
    credit_card: yup.string().matches(creditCardRegex, 'Invalid credit card number.').required()
});

// New admin validation schema.
export const adminSchema = yup.object().shape({
    first_name: customerSchema.first_name,
    last_name: customerSchema.last_name
});

// Register new user validation.
export const registerUserSchema = yup.object().shape({
    username: yup.string().min(3, "Username too short").max(60, "Username too long.").required("Username required."),
    email: yup.string().email().required("Email required"),
    password: yup.string().min(4, "Password too short").max(20, "Password too long"),
    confirmPassword: yup.string().oneOf([yup.ref('password')], "Passwords don't match")
});

// New or edit existing flight validation.
export const flightSchema = yup.object().shape({
    origin: yup.string().required("Please select flight origin country."),
    destination: yup.string().required("Please select flight destination."),
    departure_time: yup.date().min(moment().toDate(), "Invalid flight departure time.").required("Valid flight departure time required"),
    landing_time: 
        yup.date()
        .min(moment().add(30, 'minutes').toDate(), "Landing can be at least half an hour after flight departure")
        .test({
            name: 'departure',
            message: "Landing cannot be earlier than departure.",
            test: function (value) {
                return value >= this.parent.departure_time;
            }
          
        })
        .required("Valid landing time required"),
    remaining_tickets: yup.number().min(0).max(850).required("Please set the available tickets.")
});